### PR TITLE
Add integrity check

### PR DESCRIPTION
**Pull request by wijet**
Split synchronization logic into separate strategies.
Full strategy is the base one, it simply synchronizes given relation
from the API to local database, it also removes from local database
additional objects (if remove: true).

UpdatedSince strategy asks API only for changes since last sync for
given relation.

Check strategy performs integrity check, it simply fetches whole
relation from the API and displays objects which are missing, additional
or changed in the local database.
For changed objects ActiveModel #changes hash is displayed

### CODE REVIEW

File: lib/synced/strategies/check.rb [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    # with the remote ones (taken from the API).
+    class Check < Full
+      attr_reader :result
+
+      def initialize(model_class, options = {})
+        super
+        @result = Result.new
+      end
+
+      # Makes a DRY run of full synchronization. It checks and collects objects objects which
```
Comment (Mandaryn) [inline lib/synced/strategies/check.rb]: somethings wrong with this comment

File: lib/synced/strategies/full.rb [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+        end
+      end
+
+      # Returns api client from the closest possible source.
+      #
+      # @raise [BookingSync::API::Unauthorized] - On unauthorized user
+      # @return [BookingSync::API::Client] BookingSync API client
+      def api
+        return @api if @api
+        closest = [@scope, @scope.class, @model_class].detect do |o|
```
Comment (Mandaryn) [inline lib/synced/strategies/full.rb]:
```
api = [@scope, @scope.class, @model_class].detect do |o|
  o.try(:api)
end
```
no need to assign closest

File: lib/synced/strategies/full.rb [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+                    o.respond_to?(:api)
+                  end
+        closest && closest.api || raise(MissingAPIClient.new(@scope, @model_class))
+      end
+
+      def local_object_by_remote_id(remote_id)
+        local_objects.find { |l| l.public_send(@id_key) == remote_id }
```
Comment (Mandaryn) [inline lib/synced/strategies/full.rb]: decide on either `find` or `detect`

File: lib/synced/strategies/full.rb (hunk 2 above)
Comment (Fredar) [inline lib/synced/strategies/full.rb]: Better to replace single letters with actual words like `object`, `local_object` below.

File: lib/synced/strategies/full.rb [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+      def api
+        return @api if @api
+        closest = [@scope, @scope.class, @model_class].find do |object|
+                    object.respond_to?(:api)
+                  end
+        closest.try(:api) || raise(MissingAPIClient.new(@scope, @model_class))
+      end
+
+      def local_object_by_remote_id(remote_id)
+        local_objects.find { |l| l.public_send(@id_key) == remote_id }
```
Comment (Fredar) [inline lib/synced/strategies/full.rb]: And `local_object` here ;)

Comment (wijet) [inline lib/synced/strategies/full.rb]: I'm actually not convinced here, if I would made it any better using `local_object`, it's a one lines you see that a single object will be a `local_object`

**Pull request merged**